nautobot/utilities/factory.py

```diff
@@ -13,14 +46,62 @@ class Meta:
                 exclude = ("has_group,")
 
             # Required foreign key
-            user = random_instance(User)
+            user = random_instance(User, allow_null=False)
 
             # Optional foreign key
             has_group = factory.Faker("pybool")
             group = factory.Maybe("has_group", random_instance(Group), None)
+
+            # Foreign key selected from a filtered queryset
+            tenant = random_instance(Tenant.objects.filter(group__isnull=False))
+
+            # Foreign key selected from a queryset generated by a lambda
+            # This needs to be done this way because .get_for_model() evaluates a ContentType queryset,
+            # and we need to defer evaluation of that queryset as well.
+            status = random_instance(lambda: Status.objects.get_for_model(Object), allow_null=False)
+    """
+
+    def get_random_instance():
+        queryset = _get_queryset_from_model_or_queryset_or_lambda(model_or_queryset_or_lambda)
+
+        if not allow_null and not queryset.exists():
+            raise RuntimeError(f"No objects in queryset for {model_or_queryset_or_lambda}! {queryset.explain()}")
```

Comment:
Ran into an issue with `queryset.explain()` when using a django tree query queryset:

```
>>> Location.objects.all().explain()
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.SyntaxError: syntax error at or near "EXPLAIN"
LINE 26:     EXPLAIN SELECT (__tree.tree_depth) AS "tree_depth", (__t...
```

You get this when using `random_instance(Location, allow_null=False)` when no Location instances exist.

Reply:
Interesting! Sounds like maybe a bug worth reporting to django-tree-queries? Since this code is supposed to raise an exception *anyway*, do you think it's a problem that in this case we raise a `SyntaxError` instead of a `RuntimeError`?

Reply:
It looks like this may be specific to `LocationQuerySet` because `Region.objects.all().explain()` works. I'll get an issue opened to look into `LocationQuerySet`

Reply:
Region is an MPTT model, not a django-tree-queries model, so that tracks.

Reply:
Issue opened as #2577, which required fix in `django-tree-queries` which has been done in `0.12`: https://github.com/matthiask/django-tree-queries/blob/main/CHANGELOG.rst#012-2022-11-30